library/std/src/sys/unix/locks/futex.rs

```diff
@@ -162,3 +164,87 @@ impl Condvar {
         r
     }
 }
+
+/// A reentrant mutex. Used by stdout().lock() and friends.
```

Comment:
Is it intended that `stdout().lock()` is reentrant?

Reply:
Looks like it was done to avoid a deadlock in case a `Display` or `Debug` impl uses `println!()`: https://github.com/rust-lang/rust/commit/45aa6c8d1bc2f7863c92da6643de4642bb2d83bf

Reply:
> Is it intended that stdout().lock() is reentrant? 

Yup, it's the only reason we even have a ReentrantMutex in std.